shell/platform/embedder/tests/embedder_test.h

```diff
@@ -16,29 +16,63 @@
 
 namespace flutter::testing {
 
+class EmbedderTestContextGL;
+class EmbedderTestContextMetal;
+class EmbedderTestContextSoftware;
+class EmbedderTestContextVulkan;
+
 class EmbedderTest : public ThreadTest {
  public:
   EmbedderTest();
 
   std::string GetFixturesDirectory() const;
 
-  EmbedderTestContext& GetEmbedderContext(EmbedderTestContextType type);
+  template <typename T>
+  T& GetEmbedderContext() {
+    static_assert(false, "Unsupported test context type");
+  }
+
+  template <>
+  EmbedderTestContextGL& GetEmbedderContext<EmbedderTestContextGL>() {
+    return reinterpret_cast<EmbedderTestContextGL&>(GetGLContext());
+  }
+
+  template <>
+  EmbedderTestContextMetal& GetEmbedderContext<EmbedderTestContextMetal>() {
+    return reinterpret_cast<EmbedderTestContextMetal&>(GetMetalContext());
+  }
+
+  template <>
+  EmbedderTestContextSoftware&
+  GetEmbedderContext<EmbedderTestContextSoftware>() {
+    return reinterpret_cast<EmbedderTestContextSoftware&>(GetSoftwareContext());
+  }
 
- private:
-  std::map<EmbedderTestContextType, std::unique_ptr<EmbedderTestContext>>
-      embedder_contexts_;
+  template <>
+  EmbedderTestContextVulkan& GetEmbedderContext<EmbedderTestContextVulkan>() {
+    return reinterpret_cast<EmbedderTestContextVulkan&>(GetVulkanContext());
+  }
```

Comment:
Alright we still need some sneaky casts in the template specialisations here... but they're nice and hidden away.